File: cpp/Lec 2 Build-up Logical Thinking (Patterns)/pattern_15.cpp
// ABCDE
// ABCD
// ABC
// AB
// A

// BRUTE FORCE
#include <iostream>
using namespace std;
int main()
{
    int n = 5;
    for (int i = n; i >= 1; i--)
    {
        for (int j = 0; j < i; j++)
        {
            cout << char('A' + j);
        }
        cout << endl;
    }
    return 0;
}

// BETTER
#include <iostream>
#include <string>
using namespace std;
int main()
{
    int n = 5;
    string s = "ABCDE";
    for (int i = n; i >= 1; i--)
    {
        cout << s.substr(0, i) << endl;
    }
    return 0;
}

// BEST
#include <iostream>
#include <string>
using namespace std;
int main()
{
    int n = 5;
    string s = "ABCDE";
    while (!s.empty())
    {
        cout << s << '\n';
        s.pop_back();
    }
    return 0;
}

File: python/Lec 2 Build-up Logical Thinking (Patterns)/pattern_15.py
# ABCDE
# ABCD
# ABC
# AB
# A

# BRUTE FORCE
n = 5
for i in range(n, 0, -1):
    line = ""
    for j in range(i):
        line += chr(ord('A') + j)
    print(line)

# BETTER
n = 5
s = "ABCDE"
for i in range(n, 0, -1):
    print(s[:i])

# BEST
n = 5
s = "ABCDE"
while s:
    print(s)
    s = s[:-1]